Cart AJAX

coderkevin opened the pull request:
This PR adds AJAX functionality to the cart page.

Supports:
- [x] Coupons add/remove
- [x] Item Quantity
- [x] Item Removals

This completes issue #6734

coderkevin committed:
Cart ajax: apply coupon

Adds an ajax call for applying coupons while on the cart
screen (not checkout, as it already has this.)

This is the first commit to add ajax calls to the cart update
functions. See issue #6734

coderkevin committed:
Add notice to add coupon in cart.

Add the notice to when a coupon is added to the cart,
or the error if the coupon was not added for some reason.

coderkevin committed:
Block cart form while applying coupon.

Add code to block the cart form during the apply operation
until completion. Refactor block/unblock into local methods
for reuse.

coderkevin committed:
Add remove coupon ajax handling in cart.

Add code to handle ajax call for removing a coupon from the cart.

coderkevin committed:
cart.js: Consolidate url calculation.

DRY the url calculation code within a single function to make
the code more readable and succinct.

coderkevin committed:
Fix spacing.

Add space after !

coderkevin committed:
Cart AJAX: Add update call.

Since the code is built to use a form submit for all the quanitities
already, I just made this an ajax call instead of a whole page call.
The result is the exactly the same and the .woocommerce div is replaced
with the resulting HTML.

coderkevin committed:
Cart AJAX: Add cart item remove.

Add cart item remove, using same technique as the cart update.
Make the normal call via AJAX and update the HTML afterwards.

justinshreve commented on assets/js/frontend/cart.js:
```diff
@@ -6,6 +6,42 @@ jQuery( function( $ ) {
+	var update_wc_div = function( html_str ) {
+		var $html = $.parseHTML( html_str );
+		var $new_div = $( 'div.woocommerce', $html );
+		$( 'div.woocommerce').replaceWith( $new_div );
```
Missing space between single quote and end parenthesis

coderkevin replied:
Got it, thanks!

justinshreve commented on assets/js/frontend/cart.js:
```diff
@@ -17,24 +53,139 @@ jQuery( function( $ ) {
+			complete: function() {
+				update_cart_totals();
+			}
+		});
```
There should be a closing space between `}` and `)` on this and some of the other lines

coderkevin replied:
I thought so too, but the existing code did this, so I kept the pattern. Should I update all code here?

justinshreve replied:
I generally just focus on updating/fixing them when the lines are already changing (instead of making extra changes for the sake of coding standards).

coderkevin replied:
Good to know. I'll make sure to do that in the future. For this one, I can update the whole file in a single commit if we want.

claudiosmweb replied:
:+1: for Justin.
We need to remove or old bad code.

justinshreve commented on assets/js/frontend/cart.js:
```diff
@@ -17,24 +53,139 @@ jQuery( function( $ ) {
+		var $form = $( 'div.woocommerce > form' );
+
+		// Provide the submit button value because wc-form-handler expects it.
+		$( '<input />').attr( 'type', 'hidden' )
```
Space after single quote

coderkevin replied:
Got it.

justinshreve commented on assets/js/frontend/cart.js:
```diff
@@ -17,24 +53,139 @@ jQuery( function( $ ) {
+		// Provide the submit button value because wc-form-handler expects it.
+		$( '<input />').attr( 'type', 'hidden' )
+		               .attr( 'name', 'update_cart' )
+		               .attr( 'value', 'Update Cart' )
```
We will have to pass in a translated i18n string here. 
https://codex.wordpress.org/I18n_for_WordPress_Developers#Handling_JavaScript_files

coderkevin replied:
This is just to satisfy the form processing code on the server. It doesn't actually check for the value of the field, just that it exists. I don't think it would be necessary to translate this since the user never sees it and the server code doesn't care. However, we could change the value to something else if that would make it more evident. It just needs to evaluate to true in the PHP code.

justinshreve commented:
I left a few notes, mostly coding style things but one regarding an untranslated string.

coderkevin committed:
Cart AJAX: Block form during update

Apply the JQuery Block UI to the form during an AJAX update.
Also correct a few small style issues.

mikejolley commented:
LGTM :+1:

coderkevin committed:
Style: Add spaces between closing braces and parens.

claudiosmweb commented:
Great work :)

I just tested and here my thoughts:

- Need some feedback after clicking in the "x" remove button.
- When updating the cart item (by clicking on the update cart button) will be great to block the Cart Totals too (or demonstrate in some way that that panel has been updated...).

coderkevin commented:
Thanks, @claudiosmweb
After ef4debace74e84e1bef40e3fc7810cad693cda49 the form blocks upon clicking the Update Cart button or the X button for removing an item. Does this work?

claudiosmweb commented:
@conschneider no look here, still nothing shows when you click in the X button and after some minutes the products just vanish :P

Any look with the shipping form?

claudiosmweb commented:
I found a small bug too, it allows add an amount bigger than has in stock. Today we do HTML5 validation for it to check if we can update the cart.

To reproduce, you need to set any stock quantity for any product:

![screenshot from 2016-02-15 12-58-04](https://cloud.githubusercontent.com/assets/1264099/13052051/c496b2da-d3e3-11e5-81ab-9edf4aefda0a.png)

Then add your product to the card and enter a bigger number. Don't use the arrows, you need to enter the number :PYou'll see the HTML5 validation, turning the input red, but we don't check it before send the ajax request.

coderkevin commented:
@claudiosmweb Good catch on the removal. I put the code in to block the form, and I thought it was just updating too fast to see. After installing the OSX Network Link Conditioner, it looks like the block isn't appearing. I'll take a look.

coderkevin commented:
![screen shot 2016-02-15 at 9 24 44 am](https://cloud.githubusercontent.com/assets/945228/13052718/03ccf77a-d3c6-11e5-83bc-88e4304c945f.png)

@claudiosmweb Is this what you're talking about? I'll take a look.

coderkevin committed:
Cart AJAX: Fix block UI for remove item.

The JQuery selector was incorrect and needed to go
further up the DOM tree.

coderkevin commented:
@claudiosmweb Fixed the JQuery selector issue that was resulting in the form not blocking. I used the OSX Network Link Conditioner to increase the delay to confirm.

coderkevin committed:
Cart ajax: Fix html5 validation checks for quantities.

The form button click event handling was subverting the validation
checks. Moving the event to the actual form submit event fixed it.

coderkevin committed:
Cart AJAX: Clean up form handling for cart form.

The coupon and update cart logic was tripping on each other
between the click handling and form submit handling.
This commit combines the event handler, then splits off from there.

coderkevin commented:
@claudiosmweb I fixed the validation code by changing the event hook from a button click event to the actual form submit event. I had to change the form handling to make this work for both Update Cart and Apply Coupon.

coderkevin committed:
Cart AJAX: Add shipping form support.

Add support for making an AJAX call for updating the shipping options.

claudiosmweb commented:
@coderkevin great work :)
